pkg/probe/http/request.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/component-base/version"
+	"k8s.io/kubernetes/pkg/probe"
+)
+
+// NewProbeRequest returns an http.Request suitable for use as a request for a
+// probe.
+func NewProbeRequest(url *url.URL, headers http.Header) (*http.Request, error) {
+	return newProbeRequest(url, headers, "probe")
+}
+
+// NewRequestForHTTPGetAction returns an http.Request derived from httpGet.
+// When httpGet.Host is empty, podIP will be used instead.
+func NewRequestForHTTPGetAction(httpGet *v1.HTTPGetAction, container *v1.Container, podIP string, userAgentFrament string) (*http.Request, error) {
+	scheme := strings.ToLower(string(httpGet.Scheme))
+	if scheme == "" {
+		scheme = "http"
+	}
+
+	host := httpGet.Host
+	if host == "" {
+		host = podIP
+	}
+
+	port, err := probe.ResolveContainerPort(httpGet.Port, container)
+	if err != nil {
+		return nil, err
+	}
+	path := httpGet.Path
+	url := formatURL(scheme, host, port, path)
+	headers := v1HeaderToHTTPHeader(httpGet.HTTPHeaders)
+
+	return newProbeRequest(url, headers, userAgentFrament)
+}
+
+func newProbeRequest(url *url.URL, headers http.Header, userAgentFragment string) (*http.Request, error) {
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if headers == nil {
+		headers = http.Header{}
+	}
+	if _, ok := headers["User-Agent"]; !ok {
+		headers.Set("User-Agent", userAgent(userAgentFragment))
+	}
+	if _, ok := headers["Accept"]; !ok {
+		// Accept header was not defined. accept all
+		headers.Set("Accept", "*/*")
+	} else if headers.Get("Accept") == "" {
+		// Accept header was overridden but is empty. removing
+		headers.Del("Accept")
+	}
+	req.Header = headers
+	req.Host = headers.Get("Host")
+
+	return req, nil
+}
+
+func userAgent(purpose string) string {
+	v := version.Get()
+	return fmt.Sprintf("kube-%s/%s.%s", purpose, v.Major, v.Minor)
+}
+
+// formatURL formats a URL from args.  For testability.
+func formatURL(scheme string, host string, port int, path string) *url.URL {
+	u, err := url.Parse(path)
+	// Something is busted with the path, but it's too late to reject it. Pass it along as is.
+	if err != nil {
+		u = &url.URL{
+			Path: path,
+		}
+	}
```

Comment:
@tallclair can this be abused somehow to exploit things?
https://go.dev/play/p/MAwfUvMDrYv

Reply:
hmm, we are not doing validation for the path https://github.com/kubernetes/kubernetes/blob/9d85e18ec0dc09f681650d04d65e3db5237bef96/pkg/apis/core/validation/validation.go#L2785-L2800

Reply:
I can't think of anything... what did you have in mind?